core/schema/middlewares.go

```diff
@@ -0,0 +1,137 @@
+func DagOpFileMiddleware[T dagql.Typed, A any](srv *dagql.Server, fn func(ctx context.Context, self dagql.Instance[T], args A) (dagql.Instance[*core.File], error)) func(ctx context.Context, self dagql.Instance[T], args A) (dagql.Instance[*core.File], error) {
+	return func(ctx context.Context, self dagql.Instance[T], args A) (inst dagql.Instance[*core.File], err error) {
+		if _, ok := core.DagOpFromContext[core.DirectoryDagOp](ctx); ok {
+			return fn(ctx, self, args)
+		}
+
+		deps, err := extractLLBDependencies(ctx, self.Self)
+		if err != nil {
+			return inst, err
+		}
+
+		filename := "file"
+		id, err := srv.SelectID(ctx, srv.Root(),
+			dagql.Selector{
+				Field: "directory",
+			},
+			dagql.Selector{
+				Field: "withFile",
+				Args: []dagql.NamedInput{
+					{
+						Name:  "path",
+						Value: dagql.String(filename),
+					},
+					{
+						Name:  "source",
+						Value: dagql.NewID[*core.File](dagql.CurrentID(ctx).WithMetadata("", true)),
+					},
+				},
+			},
+		)
+		if err != nil {
+			return inst, err
+		}
+
+		dir, err := core.NewDirectoryDagOp(ctx, srv, id, deps)
+		if err != nil {
+			return inst, err
+		}
+		f, err := dir.File(ctx, filename)
+		if err != nil {
+			return inst, err
+		}
+		return dagql.NewInstanceForCurrentID(ctx, srv, self, f)
+	}
+}
+
+// DagOpDirectoryMiddleware caches a directory field as a buildkit operation,
+// similar to DagOpFileMiddleware.
+func DagOpDirectoryMiddleware[T dagql.Typed, A any](srv *dagql.Server, fn func(ctx context.Context, self dagql.Instance[T], args A) (dagql.Instance[*core.Directory], error)) func(ctx context.Context, self dagql.Instance[T], args A) (dagql.Instance[*core.Directory], error) {
+	return func(ctx context.Context, self dagql.Instance[T], args A) (inst dagql.Instance[*core.Directory], err error) {
+		if _, ok := core.DagOpFromContext[core.DirectoryDagOp](ctx); ok {
+			return fn(ctx, self, args)
+		}
+
+		deps, err := extractLLBDependencies(ctx, self.Self)
+		if err != nil {
+			return inst, err
+		}
+
+		id, err := srv.SelectID(ctx, srv.Root(),
+			dagql.Selector{
+				Field: "directory",
+			},
+			dagql.Selector{
+				Field: "withDirectory",
+				Args: []dagql.NamedInput{
+					{
+						Name:  "path",
+						Value: dagql.String(""),
+					},
+					{
+						Name:  "source",
+						Value: dagql.NewID[*core.Directory](dagql.CurrentID(ctx).WithMetadata("", true)),
+					},
+				},
+			},
+		)
+		if err != nil {
+			return inst, err
+		}
+
+		dir, err := core.NewDirectoryDagOp(ctx, srv, id, deps)
+		if err != nil {
+			return inst, err
+		}
+		return dagql.NewInstanceForCurrentID(ctx, srv, self, dir)
+	}
+}
+
+func extractLLBDependencies(ctx context.Context, a any) ([]llb.State, error) {
+	hasPBs, ok := a.(core.HasPBDefinitions)
+	if !ok {
+		return nil, nil
+	}
+
+	depsDefs, err := hasPBs.PBDefinitions(ctx)
+	if err != nil {
+		return nil, err
+	}
+	deps := make([]llb.State, 0, len(depsDefs))
```

Comment:
nit: it looks like some of our impls of `PBDefinitions` use a map so the order is random, which I *think might* change the buildkit cache key and result in extra invalidation?

I'm sort of doubtful this is anything remotely near a big deal atm, so not a blocker. If there's a really easy fix then maybe worth it, otherwise just worth keeping in mind